h2o-core/src/main/java/water/parser/CsvParser.java

```diff
@@ -149,21 +157,24 @@
             str.addBuff(bits);
           }
           if( !isNa &&
-              _setup.isNA(colIdx, str.toBufferedString())) {
+              _setup.isNA(columnCounter, str.toBufferedString())) {
             isNa = true;
           }
-          if (!isNa) {
-            dout.addStrCol(colIdx, str.toBufferedString());
+
+          if (!isNa && (colIdx <= colIndexNum) && _keepColumns[colIdx]) {
+            dout.addStrCol(columnCounter, str.toBufferedString());
             if (!isAllASCII)
-              dout.setIsAllASCII(colIdx, isAllASCII);
+              dout.setIsAllASCII(columnCounter, isAllASCII);
           } else {
-            dout.addInvalidCol(colIdx);
+            if ((colIdx <= colIndexNum) && _keepColumns[colIdx])
+              dout.addInvalidCol(columnCounter);
             isNa = false;
           }
           str.set(null, 0, 0);
           quotes = 0;
           isAllASCII = true;
-          ++colIdx;
+          if ((colIdx <= colIndexNum) &&  _keepColumns[colIdx++] && (columnCounter < parseIndexNum)) // only increment if not at the end
```

Comment:
Incresing `colIdx` at the end of column `c` instead of increasing it at the beginning of column `c + 1` seems like a good change.

Reply:
Agree.